### DISCUSSION TITLE
Varying performance based on slightly different data

### DISCUSSION THREAD

**Discussion by user01** [pyopencl/pyopencl#560] [Troubleshooting]
Hi,

I've been experimenting with `pyopencl` by implementing a specific function from [h3](https://github.com/uber/h3).

During my tests, I've noticed very different performance based on small changes to the provided sets (~10x). The timings are quite repeatable based on multiple executions. Also interestingly, I don't see a pattern for which sets will perform differently.

![point_sizes](https://user-images.githubusercontent.com/1885977/167331829-4b4de823-f860-4f96-af7d-a06553cf3ece.png)

This is [a gist](https://gist.github.com/user01/514a9ad748858310783c297d42659c45) with a test harness and the .cl functions I modified from uber's codebase. clinfo is attached, but in short it's a debian os with a 3090

```bash
❯ python pycl_exp.py
OpenCL for 18,001,933 in 1.48301 seconds
OpenCL for 18,001,932 in 0.15895 seconds
OpenCL for 18,001,933 in 1.35726 seconds
```

![pyopencl_profile_annotated](https://user-images.githubusercontent.com/1885977/167331321-6a100825-be3b-433f-8e77-856bfa2c83d7.png)

Would very much appreciate ideas for why the `read_buffer` would vary so wildly based on the input data. Recommendations for how to approach the task or debug the process? Am I missing something in the process?

Thanks for your time on this question and the library overall,
Erik

**Comment by inducer** [comment]
- Which ICD are you using? (Straight Nvidia or Nvidia via pocl (`conda install pocl-cuda`)?)
- What happens if you try the respective other one?

**Reply by user01** [reply]
Straight NVIDIA.

I tried `conda install pocl-cuda` but noticed no new platforms in `clinfo` or `examples/dump-properties.py`. Interestingly, no error messages. Conda claims `pocl-cuda-1.8` is installed with the transaction, but nothing appears for platforms.

I also grabbed the pocl source and tried to build it myself (`cmake -DENABLE_CUDA=ON ..` etc). The build failed with a missing clang header - perhaps wrong version of clang?

I'll give it another try when I can. Thanks for your reply and help so far

**Reply by inducer** [reply]
One thing to note is that `clEnqueueReadBuffer` (which you point to for having varying execution times) is, in principle, doing very little work: all it is supposed to do is add an operation to a queue, a quick, amortized O(1) operation.

Now, suppose that Nvidia has some "fill level" controls on their queues (not unrealistic given that they're possibly ring buffers under the hood), some of these `clEnqueueReadBuffer` may need to wait for operations at the front of the queue to complete. This may serve to explain what's going on, and on its own I think isn't necessarily indication of a problem.

The Nvidia profiler is likely to offer a better and more detailed picture of what's happening. Unfortunately, the Nvidia-provided OpenCL ICD does not permit its use, for whatever reason (only Nvidia would know). pocl-cuda allows you to use the Nvidia profiler, though, so that would be my suggestion.

**Reply by isuruf** [reply]
> I tried conda install pocl-cuda but noticed no new platforms in clinfo or examples/dump-properties.py. Interestingly, no error messages. Conda claims pocl-cuda-1.8 is installed with the transaction, but nothing appears for platforms.

You need to install pyopencl from conda as well. Did you install pyopencl with pip?
